app_dart/lib/src/request_handlers/refresh_chromebot_status.dart

```diff
@@ -80,27 +83,48 @@ class RefreshChromebotStatus extends ApiRequestHandler<Body> {
     /// [builderNumberList] when luci rerun happens, and update [devicelabTask]
     /// status when the status of latest luci run changes.
     for (FullTask datastoreTask in datastoreTasks) {
-      if (luciTasksMap.containsKey(datastoreTask.commit.sha)) {
-        final List<LuciTask> luciTasks = luciTasksMap[datastoreTask.commit.sha];
-        final String buildNumberList =
-            luciTasks.reversed.map((LuciTask luciTask) => luciTask.buildNumber.toString()).toList().join(',');
-        if (buildNumberList != datastoreTask.task.buildNumberList ||
-            luciTasks.first.status != datastoreTask.task.status) {
-          final Task update = datastoreTask.task;
-          update.status = luciTasks.first.status;
-          update.buildNumberList = buildNumberList;
-          update.builderName = builder.name;
-          update.luciBucket = 'luci.flutter.prod';
-          await datastore.insert(<Task>[update]);
-          // Save luci task record to BigQuery only when task finishes.
-          if (update.status == Task.statusFailed || update.status == Task.statusSucceeded) {
-            await _insertBigquery(update);
-          }
+      final String commitSha = datastoreTask.commit.sha;
+      if (!luciTasksMap.containsKey(commitSha)) {
+        continue;
+      }
+      final List<LuciTask> luciTasks = luciTasksMap[commitSha];
+      final String buildNumberList =
+          luciTasks.reversed.map((LuciTask luciTask) => luciTask.buildNumber.toString()).toList().join(',');
+      final LuciTask latestLuciTask = luciTasks.first;
+      if (buildNumberList != datastoreTask.task.buildNumberList || latestLuciTask.status != datastoreTask.task.status) {
+        final Task update = datastoreTask.task;
+        update.status = latestLuciTask.status;
+
+        if (_shouldRerunTask(latestLuciTask, update.attempts)) {
+          await luciBuildService.rescheduleProdBuild(
+            commitSha: commitSha,
+            builderName: latestLuciTask.builderName,
+          );
+          update.status = Task.statusNew;
+          update.attempts += 1;
+        }
+
+        update.buildNumberList = buildNumberList;
+        update.builderName = builder.name;
+        update.luciBucket = 'luci.flutter.prod';
+        await datastore.insert(<Task>[update]);
+        // Save luci task record to BigQuery only when task finishes.
+        if (update.status == Task.statusFailed || update.status == Task.statusSucceeded) {
+          await _insertBigquery(update);
```

Comment:
can we move the retry logic to a service class out of the handler.

Reply:
Moved to `luci_build_service`, so that the following up engine change can share the code. Also added more unit tests.